src/vmm/src/builder.rs

```diff
@@ -148,26 +148,372 @@ impl std::convert::From<linux_loader::cmdline::Error> for StartMicrovmError {
+        Ok(vcpus)
+    }
+
+    pub(crate) fn load_kernel(
+        boot_config: &BootConfig,
+        guest_memory: &GuestMemoryMmap,
+    ) -> Result<GuestAddress, StartMicrovmError> {
+        let mut kernel_file = boot_config
+            .kernel_file
+            .try_clone()
+            .map_err(|err| StartMicrovmError::Internal(VmmError::KernelFile(err)))?;
+
+        let entry_addr = Loader::load::<std::fs::File, GuestMemoryMmap>(
+            guest_memory,
+            None,
+            &mut kernel_file,
+            Some(GuestAddress(crate::arch::get_kernel_start())),
+        )
+        .map_err(StartMicrovmError::KernelLoader)?;
+
+        Ok(entry_addr.kernel_load)
+    }
+
+    pub(crate) fn create_vmm_and_vcpus(
+        instance_info: &InstanceInfo,
+        guest_memory: GuestMemoryMmap,
+        uffd: Option<Uffd>,
+        vm_config: &VmConfig,
+        kvm_capabilities: Vec<KvmCapability>,
+    ) -> Result<(Vmm, Vec<Vcpu>), StartMicrovmError> {
+        let mut vmm = build_vmm(
+            instance_info,
+            guest_memory,
+            uffd,
+            vm_config,
+            kvm_capabilities,
+        )?;
+
+        let pio_device_manager = {
+            // Serial device setup.
+            let serial_device = setup_serial_device(std::io::stdin(), io::stdout())
+                .map_err(StartMicrovmError::Internal)?;
+
+            // x86_64 uses the i8042 reset event as the Vmm exit event.
+            let reset_evt = vmm
+                .vcpus_exit_evt
+                .try_clone()
+                .map_err(VmmError::EventFd)
+                .map_err(StartMicrovmError::Internal)?;
+
+            setup_interrupt_controller(&mut vmm.vm).map_err(StartMicrovmError::Internal)?;
+
+            // create pio dev manager with legacy devices
+            let pio_device_manager = {
+                // TODO Remove these unwraps.
+                let mut pio_dev_mgr = PortIODeviceManager::new(serial_device, reset_evt).unwrap();
+                pio_dev_mgr.register_devices(vmm.vm.fd()).unwrap();
+                pio_dev_mgr
+            };
+
+            pio_device_manager
+        };
+
+        vmm.pio_device_manager = Some(pio_device_manager);
+
+        let vcpus = create_vcpus(&mut vmm.vm, vm_config.vcpu_count, &vmm.vcpus_exit_evt)
+            .map_err(StartMicrovmError::Internal)?;
+
+        Ok((vmm, vcpus))
+    }
+}
+
+fn build_vmm(
     instance_info: &InstanceInfo,
-    event_manager: &mut EventManager,
     guest_memory: GuestMemoryMmap,
     uffd: Option<Uffd>,
-    track_dirty_pages: bool,
-    vcpu_count: u8,
+    vm_config: &VmConfig,
     kvm_capabilities: Vec<KvmCapability>,
-) -> Result<(Vmm, Vec<Vcpu>), StartMicrovmError> {
+) -> Result<Vmm, StartMicrovmError> {
     use self::StartMicrovmError::*;
 
     // Set up Kvm Vm and register memory regions.
     // Build custom CPU config if a custom template is provided.
+    //
+    // allow unused_mut for the aarch64 platform.
+    #[allow(unused_mut)]
```

Comment:
Since this `mut` was only needed for `setup_interrupt_controller` call, it can be removed.